### ISSUE TITLE
[ppc64le] Seg fault with distributed training using multiple data loading workers

### ISSUE THREAD

**Issue by vaisaxena** [pytorch/pytorch#4934]
## Environment

    OS: RHEL 7.3
    Architecture: ppc64le
    PyTorch version: 0.4.0a0+f94f572
    How you installed PyTorch (conda, pip, source): source
    Python version: Python 3.6.2 :: Anaconda, Inc.
    CUDA/cuDNN version: 9.1/7.0.3
    GPU models and configuration: Nvidia Tesla P100
    GCC version  (if compiling from source): 4.8.5
    MPI: OpenMPI 2.1.1

## Problem

1. I ran the imagenet training example (https://github.com/pytorch/examples/tree/master/imagenet) in distributed mode (with OpenMPI) with non-zero data loading workers (>0) in torch.utils.data.DataLoader. In this case, the application either hangs or seg faults after showing following warning from OpenMPI:

[Output of mpirun  -np 4 -x LD_LIBRARY_PATH python main.py  -a resnet50 --dist-backend mpi --world-size 4 --epochs 2 -j 4 /gpfs/r92gpfs02/vaibhav/shared/small-imagenet]

creating model 'resnet50'
creating model 'resnet50'
creating model 'resnet50'
creating model 'resnet50'

--------------------------------------------------------------------------
A process has executed an operation involving a call to the
"fork()" system call to create a child process.  Open MPI is currently
operating in a condition that could result in memory corruption or
other system errors; your job may hang, crash, or produce silent
data corruption.  The use of fork() (or system() or other calls that
create child processes) is strongly discouraged.

The process that invoked fork was:

  Local host:          [[36650,1],0] (PID 144066)

If you are *absolutely sure* that your application will successfully
and correctly survive a call to fork(), you may disable this warning
by setting the mpi_warn_on_fork MCA parameter to 0.
--------------------------------------------------------------------------"

The complete stack trace of the seg fault is provided in the attached file debug_4node_j4_e2.log
[debug_4node_j4_e2.log](https://github.com/pytorch/pytorch/files/1677387/debug_4node_j4_e2.log)

2. The example works fine in serial mode (no MPI) with multiple data loading workers (running with -j > 0) 

3. The example also works fine in distributed mode (with OpenMPI) with no separate data loading worker (running the application with -j 0). In this case all the data loading happens in the main process itself. There is no "fork" related warning from OpenMPI in this case. 

So there seems to be some problem using OpenMPI with multiple data loading workers. 

I also noticed a warning in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L24 that DistributedDataParallel only works with "nccl" and "gloo", so I changed the imagenet/main.py to use torch.nn.DataParallel with a separate gradient average function instead of torch.nn.parallel.DistributedDataParallel in the distributed case (with MPI) but that also didn't help.

**Comment by vaisaxena** [comment]
Any fix or workaround for this problem? I am kind of stuck with my distributed runs because of this and any help will be appreciated.

**Comment by vaisaxena** [comment]
The problem gets resolved by changing the multiprocessing start method to 'spawn' by calling 

import torch.multiprocessing as mp
if  \_\_name__ == '\_\_main\_\_':
    mp.set_start_method('spawn') 

OpenMPI still gives fork warning as before but the code runs fine without any problem.
